server/src/main/java/com/vaadin/server/VaadinService.java

```diff
@@ -641,21 +650,27 @@ protected void lockSession(WrappedSession wrappedSession) {
             lock.unlock();
             throw e;
         }
+        return lock;
     }
 
     /**
      * Releases the lock for the given session for this service instance.
      * Typically you want to call {@link VaadinSession#unlock()} instead of this
      * method.
+     * <p>
+     * Note: Overriding this method is not recommended, for custom lock storage 
+     * strategy override {@link #getSessionLock(WrappedSession)} and
+     * {@link #setSessionLock(WrappedSession,Lock)} instead.
      *
      * @param wrappedSession
-     *            The session to unlock
+     *            The session to unlock, used only with assert
+     * @param lock
+     *            Lock instance to unlock
      */
-    protected void unlockSession(WrappedSession wrappedSession) {
+    protected void unlockSession(WrappedSession wrappedSession, Lock lock) {
         assert getSessionLock(wrappedSession) != null;
```

Comment:
This assert seems quite pointless now, and would make the application throw without unlocking if assertions are enabled in the scenario that we're trying to fix.

Reply:
Yes, sort of. If I drop it, having WrappedSession will be un-used parameter. But could be point to preserve that in case of someone has indeed overriden this method and used WrappedSession for some other purpose than obtaining the Lock instance.